app/models/concerns/encryptable.rb

```diff
@@ -1,115 +1,120 @@
-    end
-
-    def is_encryptable?(str)
-      return true if !matches_prefix?(str) && str.present?
-      if str.blank?
-        puts_and_logs "String is empty', so #{self.class.name} #{name} was not encrypted"
-      else
-        puts_and_logs "String starts with the prefix '#{ENCRYPTION_PREFIX}', so #{self.class.name} #{name} was not encrypted again"
+    def define_getter_in_db(fields)
+      fields.each do |field|
+        define_method "#{field}_in_db" do
+          read_attribute(field.to_sym)
+        end
       end
-      false
-    end
-
-    def is_decryptable?(str)
-      return true if matches_prefix?(str)
-      puts_and_logs "String does not start with the prefix '#{ENCRYPTION_PREFIX}', so #{self.class.name} #{name} was not decrypted"
-      false
     end
 
-    def encrypt_field(str)
-      return str.to_s unless is_encryptable?(str)
-      encryptor = ActiveSupport::MessageEncryptor.new(ENCRYPTION_KEY)
-      begin
-        # add prefix to encrypted string
-        str_encrypted = "#{ENCRYPTION_PREFIX}#{encryptor.encrypt_and_sign(str)}"
-        puts_and_logs "Successfully encrypted field for #{self.class.name} #{name}"
-        str = str_encrypted
-      rescue
-        puts_and_logs "WARNING: Encryption failed for string. Please check that the ENCRYPTION_KEY has not changed."
+    def define_auto_decrypt_getter(fields)
+      fields.each do |field|
+        define_method field do
+          decrypt_field(send("#{field}_in_db".to_sym))
+        end
       end
-      str
     end
+  end
 
-    def decrypt_field(str)
-      return str unless is_decryptable?(str)
-      encryptor = ActiveSupport::MessageEncryptor.new(ENCRYPTION_KEY)
-      begin
-        # remove prefix before decrypting string
-        str_no_prefix = str.gsub(/^#{ENCRYPTION_PREFIX}/, "")
-        str_decrypted = encryptor.decrypt_and_verify(str_no_prefix)
-        puts_and_logs "Successfully decrypted field for #{self.class.name} #{name}"
-        str = str_decrypted
-      rescue ActiveSupport::MessageVerifier::InvalidSignature
-        puts_and_logs "WARNING: Decryption failed for string. Please check that the ENCRYPTION_KEY has not changed."
+  def encrypt_setters
+    self.encryptable_fields.each do |field|
+      if send("#{field}_changed?")
+        self.send("#{field}=", encrypt_field(read_attribute(field.to_sym)))
       end
-      str
     end
+  end
 
-    module ClassMethods
-      def encrypts(*fields)
-        class_attribute :encryptable_fields
-        self.encryptable_fields = fields.map(&:to_sym)
-        define_getter_in_db(fields.map(&:to_sym))
-        define_auto_decrypt_getter(fields.map(&:to_sym))
-      end
+  def matches_prefix?(str)
+    ENCRYPTION_PREFIX == str.to_s[0..(ENCRYPTION_PREFIX.length - 1)]
+  end
 
-      def encrypts?(field)
-        encryptable_fields.include?(field.to_sym)
-      end
+  def encryption_key
+    return ENV['ENCRYPTION_KEY'] if ENV['ENCRYPTION_KEY'].present?
```

Comment:
This is just meant to be a simple getter, so I didn't want to log every time it was called, but instead both uses of it log appropriately (e.g. warn) if it's missing, which I think is a little clearer.

I'll remove the ENV usage from the example, I meant to try and move it into this concern instead of having it in the "configuration".